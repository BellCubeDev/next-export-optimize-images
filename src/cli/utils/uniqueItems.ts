import uniqWith from 'lodash.uniqwith'

import type { Manifest } from '../'

const uniqueItems = (items: Manifest) => uniqWith(items, (a, b) => a.output === b.output)

export default uniqueItems
