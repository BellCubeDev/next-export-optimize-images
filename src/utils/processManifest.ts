const processManifest = (manifestJson: string) =>
  manifestJson
    .trim()
    .split(/\n/g)
    .map((line) => JSON.parse(line))

export default processManifest
