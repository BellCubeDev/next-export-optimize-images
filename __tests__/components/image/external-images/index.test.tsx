/**
 * @jest-environment jsdom
 */
process.env['TEST_CONFIG_PATH'] = '__tests__/components/image/external-images/config.js'
process.env['TEST_JSON_PATH'] = '__tests__/components/image/external-images/manifest.json'

import path from 'path'

import { render, screen } from '@testing-library/react'
import fs from 'fs-extra'
import React from 'react'

import uniqueItems from '../../../../src/cli/utils/uniqueItems'
import CustomImage from '../../../../src/image'
import processManifest from '../../../../src/utils/processManifest'

const manifestPath = path.resolve(__dirname, 'manifest.json')

beforeAll(async () => {
  await fs.remove(manifestPath)
})
describe('External images', () => {
  test('Manifest.json is output correctly', () => {
    render(
      <CustomImage
        src="https://next-export-optimize-images.vercel.app/sub-path/og.png"
        width={1920}
        height={1280}
        priority
      />
    )

    const manifest = uniqueItems(processManifest(fs.readFileSync(manifestPath, 'utf-8')))
    expect(manifest).toEqual([
      {
        extension: 'webp',
        externalUrl: 'https://next-export-optimize-images.vercel.app/sub-path/og.png',
        output: '/_next/static/chunks/images/sub-path/og_1920_75.webp',
        quality: 75,
        src: '/_next/static/media/sub-path/og.png',
        width: 1920,
      },
      {
        extension: 'webp',
        externalUrl: 'https://next-export-optimize-images.vercel.app/sub-path/og.png',
        output: '/_next/static/chunks/images/sub-path/og_3840_75.webp',
        quality: 75,
        src: '/_next/static/media/sub-path/og.png',
        width: 3840,
      },
    ])
  })

  test('URLs of external images are set correctly', () => {
    render(
      <CustomImage
        src="https://next-export-optimize-images.vercel.app/sub-path/og.png"
        width={1920}
        height={1280}
        priority
      />
    )

    expect(screen.getByRole('img')).toHaveAttribute(
      'src',
      '/base-path/_next/static/chunks/images/sub-path/og_3840_75.webp'
    )
    expect(screen.getByRole('img')).toHaveAttribute(
      'srcset',
      '/base-path/_next/static/chunks/images/sub-path/og_1920_75.webp 1x, /base-path/_next/static/chunks/images/sub-path/og_3840_75.webp 2x'
    )
  })
})
