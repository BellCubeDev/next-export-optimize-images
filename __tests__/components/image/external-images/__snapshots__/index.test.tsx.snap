// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`External images Manifest.json is output correctly 1`] = `
Array [
  Object {
    "extension": "png",
    "output": "/_next/static/chunks/images/og_1920_75.png",
    "quality": 75,
    "src": "og.png",
    "width": 1920,
  },
  Object {
    "extension": "png",
    "output": "/_next/static/chunks/images/og_3840_75.png",
    "quality": 75,
    "src": "og.png",
    "width": 3840,
  },
]
`;
